[feature] Adapted k-neighbour concave hull plugin as QGIS 3 algorithm

rudivs opened the pull request:
## Description
This algorithm is an adaptation of the code used in the [QGIS Concave Hull Plugin](https://github.com/detlevn/QGIS-ConcaveHull-Plugin). The plugin includes some clustering functionality which I have not transferred. The changes I've made are mostly to make it work as a QGIS 3 algorithm, and to simplify some of the logic in `processAlgorithm`. The actual algorithm implementation is mostly unchanged.

This PR is intended to explore whether it would be a good idea to add this to master. I can add unit tests once that has been decided (and any other suggested changes are made).

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

rudivs committed:
[feature] Adapted k-neighbour concave hull plugin as QGIS 3 algorithm

Co-authored-by: Detlev Neumann <dneumann@geospatial-services.de>

rudivs committed:
Fix description of k-nearest output layer

The output layer may have multiple features  - rewording to be more general.

rudivs committed:
Fix null output for k greater than number of points in k-neighbour algorithm

- if k is greater than the number of points, use the number of points
- if k <3 is provided, use 3
- remove unused as_polygon function

luipir commented on python/plugins/processing/algs/qgis/ConcaveHull.py:
```diff
@@ -69,13 +69,13 @@ def initAlgorithm(self, config=None):
                                                         self.tr('Split multipart geometry into singleparts geometries'), defaultValue=False))
 
-        self.addParameter(QgsProcessingParameterFeatureSink(self.OUTPUT, self.tr('Concave hull'), type=QgsProcessing.TypeVectorPolygon))
+        self.addParameter(QgsProcessingParameterFeatureSink(self.OUTPUT, self.tr('Output layer'), type=QgsProcessing.TypeVectorPolygon))
```
is it important this change?

rudivs replied:
No, it was actually an accidental change (I wanted to change it in the k-nearest algorithm and changed it here by mistake). I will undo the change.

rudivs committed:
Add documentation and improve input parameter descriptions

rudivs commented:
Currently the algorithm supports any input geometry, but it calculates the concave hull by using the nodes as input points. I am not sure that users would expect this behaviour (I think they would expect the concave hull to follow the outer edges of the polygons). To demonstrate, consider the concave hull in the image below (green outline). I don't think the output has any meaning. I propose to restrict this algorithm to point inputs only, even though that technically makes it less functional.
![image](https://user-images.githubusercontent.com/1894530/44267387-4420dd80-a22e-11e8-9c7f-1d727237f55f.png)

luipir commented:
this seems more a case in which we have to avoid multipolygon or doing two steps doing concave hull for each part before...

rudivs committed:
Give a more descriptive output name

rudivs committed:
Remove unused function

luipir commented:
@rudivs please check code indentation

Gustry commented:
As written in the PR description, you need to run the "prepare-commit' script (this will fix the indentation and some other layout issues)

rudivs commented:
Thanks for the comments. I’m on leave at the moment, but will get back to this in a few days.

rudivs committed:
Fix formatting

luipir commented on python/plugins/processing/algs/qgis/KNearestConcaveHull.py:
```diff
@@ -0,0 +1,526 @@
+                    feedback.setProgress(int(current * total))
+
+                # A minimum of 3 points is necessary to proceed
+                if len(points) >= 3:
```
Hi @rudivs is not clear to me why this check "if len(points) >= 3" but not also at line 188.
Another point is that id len(points) < 3 the algorithm return an empty sink... is this the intended behaviour?

rudivs replied:
Both good points. It's not strictly necessary because the concave hull algorithm does its own checks, but this makes it easier to catch a clear error scenario early. I've added the check to the case when there is no grouping field. Will push through some commits shortly.

luipir commented on python/plugins/processing/algs/qgis/KNearestConcaveHull.py:
```diff
@@ -0,0 +1,526 @@
+                fid += 1
+
+        else:
+            points = []
```
Hi @rudivs, I'm not completly inside the logic of the alg, but is this part exactly the same as above executed "field_index >= 0"? the only change is the request construction. is it possible to avoid this redundancy?

rudivs replied:
@luipir this second section echoes the same basic logic, but the first section has extra steps for filtering features by the grouping field and then adding that as an attribute to the output polygons (which may be multiple). It might be possible to extract this as a function which has slightly different behaviour based on whether there's a grouping field set or not, but I think that might make it harder to parse.

rudivs committed:
Add tests for k-neighbour concave hull

luipir commented:
@rudivs I would really like to accept this feature before friday (feature freeze) and then fix any other implementation details later (do you agree @Gustry @nyalldawson?)... but travis have to be happy! can you fix failing test @rudivs ?

rudivs committed:
Fix failing test for K-nearest concave hull group by field

rudivs committed:
Fix k-nearest neighbour algorithm icon not showing

rudivs commented:
Tests passed! Now to make a couple of last fixes and hope it stays green...

rudivs committed:
Improve k-nearest neighbour algorithm logic and add descriptive comments

rudivs committed:
Fix indentation

rudivs commented:
Layout is passing, and I'll assume the tests pass too (will confirm in the morning). Unless there are any other comments I've made all the changes I can at this stage. I'd be happy to implement any other suggestions either before or after merge. Let me know if I should do a rebase squash or if it's okay to merge as is.

rudivs committed:
Initialise writer after appending field

nyalldawson commented on python/plugins/processing/algs/qgis/KNearestConcaveHull.py:
```diff
@@ -0,0 +1,547 @@
+                                out_feature.setAttributes([fid, unique]) # Give the polygon the same attribute as the point grouping attribute
+                                sink.addFeature(out_feature, QgsFeatureSink.FastInsert)
+                                success = True  # at least one polygon created
+                        except:
```
I'd be inclined to agree with that... better to get the exception and know what we need to fix then silently hide errors

rudivs replied:
@nyalldawson if a valid output can't be created from the given input, I'm raising a `QgsProcessingException`. Is it recommended to also use `feedback.reportError()` or is that redundant? Assuming it's redundant and leaving it out for now, but happy to add it back.

nyalldawson replied:
It's redundant -- it will result in duplicate errors

rudivs committed:
Update icon to fit QGIS theme, rename and apply to ConcaveHull also

- Based colours and sizes on mAlgorithmDelaunay.svg
- Renamed from mAlgorithmKNearestConcaveHull.svg to mAlgorithmConcaveHull.svg to be more general
- Applied to ConcaveHull.py also to indicate similarity between algorithms

rudivs commented on python/plugins/processing/algs/qgis/KNearestConcaveHull.py:
```diff
@@ -0,0 +1,547 @@
+        return 'knearestconcavehull'
+
+    def displayName(self):
+        return self.tr('Concave hull (using k-nearest neighbour algorithm)')
```
Done. Applied matching changes to the current alpha shapes `ConcaveHull.py` algorithm too. Style guide question: does QGIS have a defined US/international language? I've used the international spelling (neighbour) but if there's a US convention, I should change that to neighbor.

nyalldawson replied:
You need to use US (shudder). There's even a spell check test on Travis which will complain if you use the correct (!) spelling ;)

rudivs replied:
Ah well, QGIS is great in so many other ways, I guess I can overlook the spelling! Made the changes...

rudivs commented on python/plugins/processing/algs/qgis/KNearestConcaveHull.py:
```diff
@@ -0,0 +1,547 @@
+                    request.setSubsetOfAttributes([])
+                    features = source.getFeatures(request) # Get features with the grouping attribute equal to the current grouping value
+                    for in_feature in features:
+                        points.extend(extract_points(in_feature.geometry())) # Either points or vertices of more complex geometry
```
Good catch. Done.

rudivs commented on python/plugins/processing/algs/qgis/KNearestConcaveHull.py:
```diff
@@ -0,0 +1,547 @@
+            features = source.getFeatures(request) # Get all features
+            total = 100.0 / source.featureCount() if source.featureCount() else 0
+            for in_feature in features:
+                points.extend(extract_points(in_feature.geometry())) # Either points or vertices of more complex geometry
```
Done.

rudivs commented on python/plugins/processing/algs/qgis/KNearestConcaveHull.py:
```diff
@@ -0,0 +1,547 @@
+    :return: list of points
+    """
+    multi_geom = QgsGeometry()
+    temp_geom = []
```
Good point. There are probably some other places where we can use native implementations from QGIS, I'm just not familiar enough with the QGIS API to know what to replace. Will refactor to remove this method.

nyalldawson replied:
There is... but we can address that after freeze :)

rudivs committed:
Changes from review comments

- Shorten algorithm names (KNearestNeighbour and ConcaveHull)
- Check for feedback cancellation
- remove unnecessary try/except blocks

rudivs committed:
Use US spelling for neighbor

rudivs committed:
Improve test coverage for k-nearest neighbor concave hull

rudivs commented:
It looks like the code layout test failed because of a connection timeout. Is it possible to restart that test, or should I push through a commit to kickstart it again?

DelazJ commented:
@rudivs, it can be rerun by committers I think but you can also close and reopen the PR to trigger a new check by Travis

rudivs committed:
Fix imports for ConcaveHull

nyalldawson commented:
Can someone please merge when Travis has finished with this one?

nyalldawson commented:
In other words :sleeping:

luipir commented:
I'm taking care of it @nyalldawson... good night

luipir commented:
congrats @rudivs
